freqtrade/leverage/funding_fee.py

```diff
@@ -0,0 +1,71 @@
+from datetime import datetime, timedelta
+from typing import List
+
+import schedule
+
+from freqtrade.persistence import Trade
+
+
+class FundingFee:
+
+    trades: List[Trade]
+    begin_times = [
+        # TODO-lev: Make these UTC time
+        "23:59:45",
+        "07:59:45",
+        "15:59:45",
+    ]
+
+    def _is_time_between(self, begin_time, end_time):
+        # If check time is not given, default to current UTC time
+        check_time = datetime.utcnow().time()
+        if begin_time < end_time:
+            return check_time >= begin_time and check_time <= end_time
+        else:  # crosses midnight
+            return check_time >= begin_time or check_time <= end_time
+
+    def _apply_funding_fees(self, num_of: int = 1):
+        if num_of == 0:
+            return
+        for trade in self.trades:
+            trade.adjust_funding_fee(self._calculate(trade.amount) * num_of)
+
+    def _calculate(self, amount):
+        # TODO-futures: implement
+        # TODO-futures: Check how other exchages do it and adjust accordingly
+        # https://www.binance.com/en/support/faq/360033525031
+        # mark_price =
+        # contract_size = maybe trade.amount
+        # funding_rate =  # https://www.binance.com/en/futures/funding-history/0
+        # nominal_value = mark_price * contract_size
+        # adjustment = nominal_value * funding_rate
+        # return adjustment
+        return
+
+    def initial_funding_fee(self, amount) -> float:
+        # A funding fee interval is applied immediately if within 30s of an iterval
+        for begin_string in self.begin_times:
+            begin_time = datetime.strptime(begin_string, "%H:%M:%S")
+            end_time = (begin_time + timedelta(seconds=30))
+            if self._is_time_between(begin_time.time(), end_time.time()):
+                return self._calculate(amount)
+        return 0.0
+
+    def start(self):
+        for interval in self.begin_times:
+            schedule.every().day.at(interval).do(self._apply_funding_fees())
+
```

Comment:
@xmatthias Let me know how you would like this done. This is the only part of futures that isn't super straight forward. Binance gives you, or takes away a fee(funding fee) at 3 specific times each day (00:00, 08:00, 16:00 UTC). So to apply this fee, we would just need to run a scheduler that applies the fee at these times each day

It's explained here 

https://www.binance.com/en/blog/421499824684900382/A-Beginners-Guide-To-Funding-Rates

but what I mentioned above is essentially it, then we just take the funding rate for a coin, and put it in a formula with the trade amount.

I just want to know how you want the scheduler done

Reply:
Taking technicallities (how to) away - for a moment ...

Assume we run this every 4 hours - and it's working as we expect / calculating as it should.

The code is based on current trades - so will not include trades closed.

What happens if i sell a trade at 12:00 UTC (or any other time in between intervals) .... will binance still remove half of the expected funding fee for that trade at 16:00 (the next interval) - or will it act as if i'd have sold this at 8:00 UTC?

Reply:
> What happens if i sell a trade at 12:00 UTC (or any other time in between intervals) .... will binance still remove half of the expected funding fee for that trade at 16:00 (the next interval) - or will it act as if i'd have sold this at 8:00 UTC?

Funding fees only apply to current trades, if you enter and exit a trade between the funding fee times, then you don't pay any funding fees

An edge case might be if an order is placed on the exchange, or if your order is liquidated, and the bot crashes, and then we can't tell if it was sold before or after the funding fee interval, which I'm just realizing should be a case for interest also. If the bot is down for a long time, we would have to get the exact liquidation or order fulfillment time, in order to calculate the funding fee and interest accurately

> I'm also not 100% certain - do funding fees only apply to binance futures - or to all futures on other exchanges?

I looked it up and Kraken has it, it looks like it's a staple of Perpetual futures contracts https://en.wikipedia.org/wiki/Perpetual_futures . That sentence " Payments are periodically exchanged..." is describing a funding fee

Reply:
Kraken's explanation is way better https://support.kraken.com/hc/en-us/articles/360022835911-Perpetual-Futures-Contract-Specifications

Reply:
if you shutdown the bot with an order on the exchange (order created by freqtrade), and this order fills while the bot is down - then freqtrade will realize this (and also get the correct fill-date - at least from binance).

That's a specific process that runs at startup (checking all open orders) - so i guess we'll have to call "something" for the funding stuff as well in that case.

seems like kraken uses a similar timing (every 4 hours ...) ... ?

Reply:
Yeah, I'm also just realizing that we need to charge interest for the entire time that an order is placed, plus the time in the trade(for margin orders, not futures, because on futures you don't borrow anything), so we'd be charging for the extra time that you are waiting for an order to fill because this is the whole time that the currency is borrowed.

Reply:
> That's a specific process that runs at startup (checking all open orders) - so i guess we'll have to call "something" for the funding stuff as well in that case.

Where does this happen? There's a funding fee method to do this, I just need to know where it should be called

Reply:
either `startup()` or `update_open_orders()` (called from startup).

It's only applying to open orders for the moment - but i think that'll be fine (we'll need to call it for still open trades anyway.